File: vol1/ch5/LeetCode379.py
from collections import deque
class PhoneDirectory:
    def __init__(self, maxNumbers: int):
        self.hset=set()
        self.qu=deque()
        for i in range(0,maxNumbers):
            self.qu.append(i)

    def get(self) -> int:
        if self.qu:
            x=self.qu.pop()
            self.hset.add(x)
            return x
        else: return -1

    def check(self, number: int) -> bool:
        if number in self.hset:
            return False
        else:
            return True

    def release(self, number: int) -> None:
        if number in self.hset:
            self.hset.remove(number)
            self.qu.append(number)

File: vol1/ch5/LeetCode379.cpp
class PhoneDirectory {
    unordered_set<int> hset;       //用于判重
    queue<int> qu;                      //用户队列
public:
    PhoneDirectory(int maxNumbers) {    //初始化
        for(int i=0;i<maxNumbers;i++)
            qu.push(i);
    }
    
    int get() {                         //分配
        if(!qu.empty()){
            int x=qu.front(); qu.pop();
            hset.insert(x);
            return x;
        }
        else return -1;
    }
    
    bool check(int number) {            //检查
        if(hset.count(number)==1)
            return false;
        else
            return true;
    }
    
    void release(int number) {          //释放
        if(hset.count(number)==1) {
            hset.erase(number);
            qu.push(number);
        }
    }
};

